File: GaestehausWild/Model/Stay.swift
import Foundation

struct Stay: Codable, Hashable, Sendable {
    var arrival: Date
    var departure: Date
    var roomID: String?
}

enum StayPhase: Hashable, Sendable {
    case notSet
    case upcoming(daysUntilArrival: Int)
    case arrivalDay
    case inHouse(nightsRemaining: Int)
    case departureDay
    case past(daysSinceDeparture: Int)
}

enum StayClock {
    static let calendar: Calendar = {
        var calendar = Calendar(identifier: .gregorian)
        calendar.locale = Locale(identifier: "de_DE")
        calendar.timeZone = TimeZone(identifier: "Europe/Berlin")!
        return calendar
    }()

    static func startOfDay(_ date: Date) -> Date {
        calendar.startOfDay(for: date)
    }

    static func nights(from start: Date, to end: Date) -> Int {
        calendar.dateComponents(
            [.day],
            from: startOfDay(start),
            to: startOfDay(end)
        ).day ?? 0
    }

    static func phase(for stay: Stay?, now: Date) -> StayPhase {
        guard let stay else { return .notSet }
        let today = startOfDay(now)
        let arrival = startOfDay(stay.arrival)
        let departure = startOfDay(stay.departure)

        if today < arrival {
            return .upcoming(daysUntilArrival: max(0, nights(from: today, to: arrival)))
        }
        if today == arrival { return .arrivalDay }
        if today < departure {
            return .inHouse(nightsRemaining: max(1, nights(from: today, to: departure)))
        }
        if today == departure { return .departureDay }
        return .past(daysSinceDeparture: max(0, nights(from: departure, to: today)))
    }

    static func isStale(_ stay: Stay, now: Date) -> Bool {
        guard startOfDay(now) > startOfDay(stay.departure) else { return false }
        return nights(from: stay.departure, to: now) > 14
    }

    static func checkIn(on day: Date) -> Date {
        calendar.date(bySettingHour: 15, minute: 0, second: 0, of: day)!
    }

    static func checkOut(on day: Date) -> Date {
        calendar.date(bySettingHour: 11, minute: 0, second: 0, of: day)!
    }
}

File: web/src/lib/stay.ts
export type Stay = { arrival: string; departure: string; roomId?: string };

export type StayPhase =
  | { kind: 'notSet' }
  | { kind: 'upcoming'; daysUntilArrival: number }
  | { kind: 'arrivalDay' }
  | { kind: 'inHouse'; nightsRemaining: number }
  | { kind: 'departureDay' }
  | { kind: 'past'; daysSinceDeparture: number };

const DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/;

export function isDateString(value: unknown): value is string {
  if (typeof value !== 'string' || !DATE_PATTERN.test(value)) return false;
  const [year, month, day] = value.split('-').map(Number);
  const date = new Date(Date.UTC(year, month - 1, day));
  return date.getUTCFullYear() === year && date.getUTCMonth() === month - 1 && date.getUTCDate() === day;
}

function ordinal(value: string): number {
  if (!isDateString(value)) throw new Error(`Invalid calendar date: ${value}`);
  const [year, month, day] = value.split('-').map(Number);
  return Date.UTC(year, month - 1, day) / 86_400_000;
}

export function addDays(value: string, days: number): string {
  const date = new Date((ordinal(value) + days) * 86_400_000);
  return date.toISOString().slice(0, 10);
}

export function nights(from: string, to: string): number {
  return ordinal(to) - ordinal(from);
}

export function berlinToday(now = new Date()): string {
  const parts = new Intl.DateTimeFormat('en-CA', {
    timeZone: 'Europe/Berlin', year: 'numeric', month: '2-digit', day: '2-digit',
  }).formatToParts(now);
  const find = (type: Intl.DateTimeFormatPartTypes) => parts.find((part) => part.type === type)?.value ?? '';
  return `${find('year')}-${find('month')}-${find('day')}`;
}

export function phaseFor(stay: Stay | null, today = berlinToday()): StayPhase {
  if (!stay) return { kind: 'notSet' };
  if (today < stay.arrival) return { kind: 'upcoming', daysUntilArrival: Math.max(0, nights(today, stay.arrival)) };
  if (today === stay.arrival) return { kind: 'arrivalDay' };
  if (today < stay.departure) return { kind: 'inHouse', nightsRemaining: Math.max(1, nights(today, stay.departure)) };
  if (today === stay.departure) return { kind: 'departureDay' };
  return { kind: 'past', daysSinceDeparture: Math.max(0, nights(stay.departure, today)) };
}

export function isStale(stay: Stay, today = berlinToday()): boolean {
  return today > stay.departure && nights(stay.departure, today) > 14;
}

export function normalizeStay(stay: Stay): Stay {
  if (!isDateString(stay.arrival) || !isDateString(stay.departure)) throw new Error('Invalid stay dates');
  return { ...stay, departure: stay.departure <= stay.arrival ? addDays(stay.arrival, 1) : stay.departure };
}

export function readStay(storage: Pick<Storage, 'getItem' | 'removeItem'>): Stay | null {
  try {
    const raw = storage.getItem('gaestehaus-wild.stay');
    if (!raw) return null;
    const parsed = JSON.parse(raw) as Stay;
    const stay = normalizeStay(parsed);
    if (isStale(stay)) {
      storage.removeItem('gaestehaus-wild.stay');
      return null;
    }
    return stay;
  } catch {
    storage.removeItem('gaestehaus-wild.stay');
    return null;
  }
}
